### PR TITLE
apply fixes to make codeoracle work

### PR DESCRIPTION
**Pull request by StanislavBreadless**
# What ❔

Ensure that reference impl supports CodeOracle & generates correct witness. Also, includes `fmt`

## Checklist

- [ ] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.
- [ ] Code has been formatted via `zk fmt` and `zk lint`.

### CODE REVIEW

File: src/opcodes/execution/log.rs [hunk 1]
```diff
@@ -167,11 +167,16 @@ impl<const N: usize, E: VmEncodingMode<N>> DecodedOpcode<N, E> {
                     is_service: is_first_message,
                 };
 
-                let refund = vm_state.refund_for_partial_query(
-                    vm_state.local_state.monotonic_cycle_counter,
-                    &partial_query,
-                );
-                let ergs_refund = refund.refund();
+                let ergs_refund = if variant == LogOpcode::StorageRead {
```
Comment (shamatar) [inline src/opcodes/execution/log.rs]: That was a whole point, that we always ask for some value, same as in circuit, and in circuit we require refund to be 0, but the number of accesses to oracle must match

Comment (shamatar) [inline src/opcodes/execution/log.rs]: Remove a branch here, add assert to 0 for transient, and hide any manipulation with oracles and witness tracer under "refund query function". Same below

**Review by shamatar — CHANGES_REQUESTED** [review]

**Comment by StanislavBreadless** [comment]
@shamatar the fact that ergs refund for tstore is 0 is enforced in the old code:

```
                if variant == LogOpcode::TransientStorageWrite {
                    assert_eq!(pubdata_cost.0, 0);
                    assert_eq!(ergs_refund, 0);
                } else {
 ```

File: src/reference_impls/memory.rs [hunk 2]
```diff
@@ -552,14 +553,29 @@ impl Memory for SimpleMemory {
...
         &mut self,
-        _monotonic_cycle_counter: u32,
-        query: MemoryQuery,
+        monotonic_cycle_counter: u32,
+        mut query: MemoryQuery,
     ) -> MemoryQuery {
         assert_eq!(query.location.memory_type, MemoryType::Code);
         let page = query.location.page.0;
 
+        // There are two types of places where decommitment could happen:
```
Comment (shamatar) [inline src/reference_impls/memory.rs:562]: Actually you do not need to check heap pages here unless we have a wrong implementation somewhere above. EVM simulator can only access opcode via uma.fat_ptr_read and this function is used only for VM opcodes

Comment (StanislavBreadless) [reply src/reference_impls/memory.rs:562]: but decommitter actually uses it to write to the memory of the code oracle

**Review by shamatar — APPROVED** [review]

**Pull request merged**